Libraries/LLM/Evaluate.swift

```diff
@@ -78,54 +34,186 @@ public struct GenerateParameters: Sendable {
+    let sampleContext: SampleContext
 
     public init(prompt: MLXArray, model: LLMModel, parameters: GenerateParameters) {
         self.model = model
         self.parameters = parameters
         self.y = prompt
-        self.cache = []
-        if parameters.repetitionContextSize > 1 {
-            if prompt.shape[0] <= parameters.repetitionContextSize {
-                self.repetitionContext = prompt
-            } else {
-                self.repetitionContext = prompt[(-parameters.repetitionContextSize)...]
-            }
-        } else {
-            self.repetitionContext = []
+        self.cache = model.newCache(parameters: parameters)
+
+        self.repetitionContext = RepetitionContext(prompt: prompt, parameters: parameters)
+        self.sampleContext = SampleContext(parameters: parameters)
+
+        // prepare the prompt in chunks if larger than the prefill size
+        while y.size > parameters.prefillStepSize {
+            _ = model(
+                y[..<parameters.prefillStepSize, .newAxis], cache: cache.isEmpty ? nil : cache)
+            eval(cache)
+            y = y[parameters.prefillStepSize...]
         }
+
+        // evaluate the remainder of the prompt -- this primes the pump
+        y = step(previous: y)
+        asyncEval(y)
     }
 
-    mutating public func next() -> MLXArray? {
+    /// Evaluate the next token and return the new token (y) and cache state.
+    ///
+    /// This may mutate the repititionContext.
+    mutating func step(previous: MLXArray) -> MLXArray {
         var logits: MLXArray
-        (logits, cache) = model(expandedDimensions(y, axis: 0), cache: cache.isEmpty ? nil : cache)
+        logits = model(previous[.newAxis], cache: cache.isEmpty ? nil : cache)
+
         logits = logits[0..., -1, 0...]
-        if let repetitionPenalty = parameters.repetitionPenalty {
-            // apply repetition penalty
-            logits = applyRepetitionPenalty(
-                logits: logits, repetitionContext: repetitionContext,
-                penalty: repetitionPenalty)
-        }
-        y = sample(logits: logits, temp: parameters.temperature, topP: parameters.topP)
-        // append the current token to the context and check repetitionPenalty context see if need to remove the first token
-        if parameters.repetitionContextSize > 1 {
-            if repetitionContext.shape[0] > parameters.repetitionContextSize {
-                repetitionContext = repetitionContext[(-parameters.repetitionContextSize)...]
-            }
-        }
+        logits = repetitionContext.applyRepetitionPenalty(logits: logits)
+
+        let y = sampleContext.sample(logits: logits)
+
+        repetitionContext.append(token: y)
 
         return y
     }
+
+    mutating public func next() -> MLXArray? {
+        // save current value -- this will be returned
+        let previousY = y
+
+        // compute the next state and async eval the next token
+        y = step(previous: previousY)
+        asyncEval(y)
```

Comment:
This restructuring supports async eval now.  We generate the graph for the next token and start the eval while returning the previously generated token.

Reply:
For comparison here is the python version:

```python
    while True:
        next_y, next_logprobs = _step(y)
        mx.async_eval(next_y)
        yield y.item(), logprobs
        y, logprobs = next_y, next_logprobs
```

This is iterator style so we are not yielding, so I restructured it a bit to keep it clear.

Not for this PR, but now that we can easily (memory-wise) evaluate off the main thread we could change this to an `AsyncStream`, which does have `yield`.  IIRC there were some issues around cleanly shutting down, but I think those can be resolved.